Entity Relations (#18)
bevyengine/rfcs: rfcs/min-relations.md

# Feature Name: `min-relations`

// FIXME
// Make sure to talk about these:
// - Reverse relations.
// - Or support for add_target_filter.

## Summary

> Relations connect entities to each other, storing data in these edges. This RFC covers the minimal version of this feature, allowing for basic entity grouping and simple game logic extensions.

Reviewer: ```suggestion
Relations allow creating logical connection between entity to multiple other entities. They also allow storing data associated with those individual connections, known as edges. Queries areexpanded to allow efficient filtering of entities that relate to specified group of entities.

This RFC covers the minimal version of this feature, allowing for basic dynamic entity grouping and simple game logic extensions.
```

Reviewer 2: Relations allow users to create logical connection between entities. Each source entity can have multiple kinds of relations, each with their own data, and multiple relations of the same kind pointing to a different target entity.
Queries are expanded to allow efficient filtering of entities by the relations they have or the targets of these relations.

## Motivation

> Entities often need to be aware of each other in complex ways: an attack that targets another unit, groups of entities that are controlled en-masse, or complex parent-child hierarchies that need to move together.

Reviewer: I think the `InFrustum(camera_id)` would be especially understandable here. It's easy to see why it can't be handled by simple tag component here.

Reviewer 2: I worry that multiple camera frustums are bit too obscure of a topic for the average gameplay programmer referring to this RFC. I agree that they're a clear example though!

We *could* simply store an `Entity` in a component, then use `query.get(my_entity)` to access the relevant data.
(In this RFC, we call this the **Entity-in-component pattern**.)
But this quickly starts to balloon in complexity.

- How do we ensure that these components are cleaned up when the entity they're pointing to is?
- How do we handle pointing to multiple entities in similar ways?
- How do we look for all entities that point to a particular entity of interest?
- How do we despawn an entity and all of its children (with many types of child-like entities)?
- How do we quickly access data on the entity that we're attached to?
- How do we traverse this graph of entities?

> - How do we ensure that our graph is acylic?

Reviewer: I'm not sure if this particular point is handled in any way by proposed relations. Afaict cycles are still an issue (in cases where you explicitly don't want them).

Reviewer 2: The framing could be clearer. These problems are *possible* to address in a cohesive way at an engine-level by including relations as a feature. Min-relations won't hit all of these though :(

Reviewer: Okay, so it's more of a future prospect. I guess that's fine.

Reviewer 2: Clarified!

We *could* force our users to solve these problems, over and over again in each individual game.

This powerful and common pattern deserves a proper abstraction that we can make ergonomic, performant and *fearless*.

## Guide-level explanation

> **Relations** are components that connect entities together, pointing from the entity they are stored on to the entity that they target.
> Just like other components, they can store effectively arbitrary data, or they can be data-less unit structs that serve as **markers**.

Reviewer: ```suggestion
**Relations** are similar to components that connect entities together, pointing from the entity they are stored on to the entity that they target.
Just like components, they can store effectively arbitrary data, or they can be data-less unit structs that serve as **markers**. Unlike components, they don't store the target entity directly, as that connection is managed by the engine.
```

    elphaba: Res<Elphaba>){
    
    // You can directly chain this as set_relation_filters returns &mut Query
    query.set_relation_filters(
        // Note that we can set relation filters even if the relation is in a query filter
        RelationFilters::new()
            // Note: add_target_filter is additive- i.e. entities this query
            // matches must be friends with *both* dorothy AND elphaba
            .add_target_filter::<FriendsWith, _>(dorothy.entity)
            .add_target_filter::<FriendsWith, _>(elphaba.entity)
    ).for_each(|mut stress| {
        println!("{} is friends with Dorothy and Elphaba!", source_entity);
        stress.val = 1000; 
    })
}

/// If we want to combine targets in an "or" fashion, we need a slightly different syntax
TODO: write example

// Query filters work too!
fn not_alone(
    commands: mut Commands, 
    query: Query<Entity, Without<Relation<FriendsWith>>>,
    puppy: Res<Puppy>
) {
    for lonely_entity in query.iter(){
        commands.insert_relation(FriendsWith, puppy.entity);
    }
}

// So does change detection with Added and Changed!
fn new_friends(query: Query<&mut Excitement, Added<Relation<FriendsWith>>>){

>     query.for_each(|mut excitement|{excitement.value += 10});

Reviewer: ```suggestion
    query.for_each(|(mut excitement, friends)|{
    	excitement.value += 10 * friends.count();
	});
```

TODO: mention relation cleanup on despawn.

Sometimes, we may instead want to look for "entities that are targets of some relation".
We can do so by iterating over all entities with that relation, and then filtering by target.

```rust
fn all_children(
    player_query: Query<Entity, With<Player>>,

>     child_query: Query<&Name, (With<Relation<ChildOf>, Without<Player>)>>,

Reviewer 2: ```suggestion
    child_query: Query<&Name, With<Relation<ChildOf>>>,
```

> fn all_children(

Reviewer: This example is equivalent to `friend_of_dorothy` and could probably be removed.

Reviewer 2: I think I agree. We were originally going to use this to demonstrate reverse relations, but we cut those from min_relations.

Reviewer 2: Removed.

    let player_entity = player_query.single.unwrap();
    
    child_query.set_relation_filters(
        RelationFilters::new()
            .add_target_filter::<ChildOf, _>(player_entity)
    );

    for name in child_query.iter() {
        println!("{} is one of my children.", name) 			
    }
}
```

Finally, you can use the `Entity` returned by your relations to fetch data from the target's components by combining it with `query::get()` or `query::get_component<C>()`.

```rust
// We need to use a marker component to avoid having multiple mutable references

> fn debts_come_due(

Reviewer 2: Replace with:
fn debts_come_due(
    mut debt_query: Query<Relation<Owes>>, 
    mut money_query: Query<&mut Money>,
) {}

Courtesy of @TheRawMeatball.

```rust

> // We need to use a marker component to avoid having multiple mutable references

Reviewer: There is no marker component in the example.

Reviewer 2: Yeah; Meatball fixed this example for us! We can remove that comment now.

Reviewer 2: Fixed.

fn debts_come_due(
    mut debt_query: Query<(Entity, Relation<Owes>)>, 
    mut money_query: Query<&mut Money>,
) {
    for (debtor, debt) in debt.query() {
        for (lender, amount_owed) in debt {
            *money_query.get(debtor).unwrap()
                .checked_sub(amount_owed)
                .expect("Nice kneecaps you got there...");
            *money_query.get(lender).unwrap() += amount_owed;
        }
    }
}
```

### Grouping entities

### Entity hierarchies

## Reference-level explanation

TODO: Boxy explains the magic.

## Drawbacks

> 1. Introduces another core abstraction to users that they will want to learn.

Reviewer: ```suggestion
1. Introduces another core abstraction to users that they will have to learn.
```

I think it was a little too optimistic :P

Reviewer 2: Sadly you're right.

Rewriting our parent-child code in terms of relations is important, but will introduce the concept to users quite quickly.

> 2. Even if relations are more ergonomic than the Entity-in-component approach to accomplish the same tasks, relations will be held to a higher bar for quality.

Reviewer: Is this a drawback? Easier and higher quality at the same time sounds like a very good thing.

Reviewer 2: Hmmm... It's more work on the devs though! I agree though, this can probably be cut.

Reviewer 2: Cut.

3. Heavy use of relations can seriously fragment archetypes. We need to be careful that this doesn't badly impact our performance.

## Rationale and alternatives

- Why is this design the best in the space of possible designs?
- What other designs have been considered and what is the rationale for not choosing them?
- What is the impact of not doing this?
- Why is this important to implement as a feature of Bevy itself, rather than an ecosystem crate?

## Prior art

[`Flecs`](https://github.com/SanderMertens/flecs), an advanced C++ ECS framework, has a similar feature, which they call "relationships".
These are somewhat different, they use an elaborate query domain-specific language along with being more feature rich.
You read more about them in the [corresponding PR](https://github.com/SanderMertens/flecs/pull/358).

## Unresolved questions

1. How do we put relations in Bundles?

> 2. What do we actually call `Noitaler`? Candidates so far are:

Reviewer: ```suggestion
2. What do we actually call [`Noitaler`][[bikeshed]? Candidates so far are:
```

   4. `TargetOf`
3. Do we need a full graph constraint solver in our queries to handle things like "check for unrequited love"?
4. Do we want a macro to make complex calls easier?

```rust
macro_rules Relation {
  (this $kind:ty *) => Relation<$kind>,
  (* $kind:ty this) => Noitaler<$kind>,
}
```

## Future possibilities

Relation enhancements:

1. Sugar for accessing data on the target entity in a single query. `Relation<T: Component, Q: WorldQuery>`
2. Graph traversals.
3. Arbitrary target types instead of just `Entity` combined with `KindedEntity` proposal to ensure your targets are valid.
4. Automatically symmetric relations.
5. `Noitaler`*
6. Graph shape guarantees using archetype invariants.
7. Streaming iters to allow for queries like: `Query<&mut Money, Relation<Owes, &mut Money>>` and potentially use invariants on graph shape to allow for skipping soundness checks for the aliasing `&mut Money`'s
8. Assorted Performance optimizations.
9. Relation orders to enable things like `Styles`.
10. Generalized `despawn_recursive`.

Relation applications in the engine:

1. Replace existing parent-child.
2. Applications for UI: styling, widget wiring etc.
3. Frustum culling.

> AUTHOR'S NOTE: Much like `yeet`, `Noitaler` is the bikeshed-avoidance name for the inverse of `Relation`, and will never be used.

Reviewer: ```suggestion
[bikeshed]: AUTHOR'S NOTE: Much like `yeet`, `Noitaler` is the bikeshed-avoidance name for the inverse of `Relation`, and will never be used.
```

struct Owes(i32);

// Let's make some entities
let world = World::new();
let alice = world.spawn().id();
let boxy = world.spawn().id();

// You can make relations with commands!
commands.entity(alice).insert_relation(FriendsWith, boxy); // uwu:3

// You can remove them with commands!
commands.entity(alice).remove_relation(FriendsWith, boxy) // not uwu :(

// You can use direct World access!
world.entity_mut(alice).insert_relation(FriendsWith, boxy); // uwu:3

// Relations are directed; a relation in one direction does not imply the same in the other direction
commands.entity(boxy).insert_relation(FriendsWith, alice); // one way friendship sad :(

// You can mix different relation kinds on the same entity!
commands.entity(alice).insert_relation(Owes(9999), boxy); // :)))))

// You can add mulitple relations of the same kind to a single source entity!
let cart  = world.spawn().id();
commands.entity(alice).insert_relation(FriendsWith, cart); // Hi!
```

Once your relations are in place, you'll want to access them in some form.
Just like with ordinary components, you can request them in your queries:

```rust
// Relations can be queried for immutably

> fn friendship_is_magic(mut query: Query<(&mut Magic, &Relation<FriendsWith>), With<Pony>>) {

Reviewer: Currently reference is necessary in query directly, but not used in combinators like `With<Relation<T>>`. It is used to determine if access is read-only or read-write. This is similar to how references are used on component types here.

    // Each entity can have their own friends
    for (mut magic, friends) in query.iter_mut(){
        // Relations return an iterator when unpacked
        magic.power += friends.count();
    }
}

// Or you can request mutable access to modify the relation's data
fn interest(query: Query<&mut Relation<Owes>>){
    for debts in query.iter(){
        for (owed_to, amount_owed) in debts {
            println!("we owe {} some money", owed_to);
            amount_owed *= 1.05; 
        }
    }
}

// You can query for entities that target a specific entity
fn friend_of_dorothy(
    mut query: Query<Entity, With<Relation<FriendsWith>>>, 
    dorothy: Res<Dorothy>,
) {
    // Setting relation filters requires mutable access to the query
    query.filter_relation::<FriendsWith, _>(
        RelationFilter::target(dorothy.entity)
    // .build() causes the relation filters to be applied

>     ).build();

Reviewer: Why is builder pattern necessary here? If we do that, then the builder should return something.

```rust
let mut filtered_query = query.filter_relation(..).build();
```

Otherwise, adding more filters would be possible after building. And also querying things without calling `build` would lead to bugs, as the filters would be not applied.

Reviewer 2: @BoxyUwU was explaining to me that the builder pattern was a necessary API sacrifice for performance in order to... something about archetype iteration? I'll add a to-do in rationale.

I think I agree with the comments about the return value though; I'll make that change.

Reviewer 2: Oh hey, using assignment allows us to avoid needing mutable access *and* eliminates the need for a reset_filter API.

I'm in favor of this.

    for source_entity in query.iter(){
        println!("{} is friends with Dorothy!", source_entity);
    }

    // RelationFilters are reset at the end of the system, make sure to set them every time your system runs!
    // You can also reset this manually by applying a new `RelationFilters` to the query
    // They override existing relation filters, rather than adding to them
}

// Or even look for some combination of targets!
fn caught_in_the_middle(mut query: Query<&mut Stress, With<Relation<FriendsWith>>>,
    dorothy: Res<Dorothy>,
    elphaba: Res<Elphaba>){

    // Note that we can set relation filters even if the relation is in a query filter
    query.filter_relation::<FriendsWith, _>(
        RelationFilter::any_of()
        .target(dorothy.entity)
        .target(elphaba.entity)
    // You can directly chain this as set_relation_filters returns &mut Query
    ).build().for_each(|mut stress| {

>         println!("{} is friends with Dorothy and Elphaba!", source_entity);

Reviewer: Yep, I believe this should be `RelationFilter::all_of()` to make sense in this example.

>         println!("{} is friends with Dorothy and Elphaba!", source_entity);

Reviewer 2: Fixed!

        stress.val = 1000; 
    })
}

// Query filters work too!
fn not_alone(
    commands: mut Commands, 
    query: Query<Entity, Without<Relation<FriendsWith>>>,
    puppy: Res<Puppy>
) {
    for lonely_entity in query.iter(){
        commands.insert_relation(FriendsWith, puppy.entity);
    }
}

// So does change detection with Added and Changed!
fn new_friends(query: Query<&mut Excitement, Added<Relation<FriendsWith>>>){
    query.for_each(|(mut excitement, new_friends)| {
        excitement.value += 10 * new_friends.count();
    });
}

```

You can use the `Entity` returned by your relations to fetch data from the target's components by combining it with `query::get()` or `query::get_component<C>()`.

```rust
fn debts_come_due(

>     mut debt_query: Query<(Entity, Relation<Owes>)>, 

Reviewer: ```suggestion
    mut debt_query: Query<(Entity, &Relation<Owes>)>, 
```

    mut money_query: Query<&mut Money>,
) {
    for (debtor, debt) in debt.query() {
        for (lender, amount_owed) in debt {
            *money_query.get(debtor).unwrap()
                .checked_sub(amount_owed)
                .expect("Nice kneecaps you got there...");
            *money_query.get(lender).unwrap() += amount_owed;
        }
    }
}
```

### Nuances of relations

Relations between entities are automatically cleaned up when either the source or target entity is removed.

> Just like components, you can view which relations were removed in the last frame through the `RemovedRelation<T>` system parameter, which also returns their data.

Reviewer: Why is `RemovedRelation` a separate thing? Should we use `Removed<Relation<T>>` instead?

Reviewer 2: I would prefer this, but I'm following the convention from components.

Reviewer: I see, it's not a query combinator, but a system param. Then the convention really should be `RemovedRelations<T>`, which probably would be an iterator over `(Entity, RemovedTargets)` where `RemovedTargets` is another iterator over `Entity` (or whatever the associated Target type will be in the future).

Reviewer 2: Ah, you're right there. I'll add the `s`. Down the line, I'd prefer to see Removed* move to subworlds, but that's out of scope here.

Each relation kind + target combination is treated like a unique type for the purposes of archetypes.
This makes can be relevant to the performance of your game; accelerating target entity filtering but creating overhead if you have a very large number of distinct targets.
As a result, changing the source or target of a relation can only be done while the app has exclusive access to the world, via `Commands` or in exclusive systems.
Here's an example of how you might do so:

```rust
fn love_potion(
    commands: mut Commands,

>     query: Query<Entity, (With<Relation<&Loves>>, With<Cute>)>,

Reviewer: That looks wrong indeed.
```suggestion
    query: Query<Entity, (With<Relation<Loves>>, With<Cute>)>,
```

    player_query: Query<Entity, With<Player>>){
    
    let player = player.single().unwrap();

    for victim in query.iter(){
        // change_target() and change_source() preserve the relation's data
        // Only one relation of a given kind can exist between a source and a target;
        // these command applies to any relations that match,
        // and overwrite any conflicting relations
        commands.entity(victim).change_target::<Loves>(player); // This is unethical!!
    }
}

fn adoption(
    commands: mut Commands,
    // As with components, you can query for relations that may or may not exist

>     query: Query<(Entity, &NewOwner, Option<Relation<Owns>>), With<Kitten>>,

Reviewer: ```suggestion
    query: Query<(Entity, &NewOwner, Option<&Relation<Owns>>), With<Kitten>>,
```

) {
   for (kitten, new_owner, previous_ownership) in query.iter(){
       // Changing the target or the source will fail silently if no appropriate relation exists
       match previous_ownership {
           // The first parameter is the data, the second is the old target
           Some(_,_) => commands.entity(kitten).change_source::<Owns>(new_owner); // uwu :3

>            None => commands.entity(kitten).insert_relation(Owns::default(), new_owner); // uwu!!

Reviewer: I'm not sure about this one, but is the relation reversed here?
```suggestion
           None => commands.entity(new_owner).insert_relation(Owns::default(), kitten); // uwu!!
```

Reviewer 2: Right, the relation should be owned by...

Reviewer 2: You're right; making the fix.

       }
   } 
}
```

### Advanced relation filters

> As the savvy reader may have guessed, we can combine relation filters in arbitrarily complex ways.
> While most of you may never need to plumb the dizzying heights of compound relation filtering,
> the DBAs in the audience will surely both thank us for our efforts and deride our naive implementations.
> 
> Fortunately, there are several simple and convenient API extensions that you might want to use.
> The first of these is the ability to filter by the source entity as well:

Reviewer: I'm not sure this whole paragraph makes much sense in the RCF concext. This is imho a bit too blog-like for techical documentation.

```suggestion
For those cases where previous examples weren't enough, there are several additional API extensions that you might want to use. Using those, Relation filters can be combined in arbitrarily complex ways.

The first of these is the ability to filter by the source entity as well:

```

```rust
fn 

```

Next, you may wish to operate over entire groups of entities at once:

```rust
fn 

```

From time-to-time, we may care about *excluding* entities who have relations to certain targets:

```rust
fn blackball(){

}

```

`any_of` and `all_of` can be used to collect groups of entities into a single filter.
`any_of` uses **or semantics**, returning any entity if any of the filters are met,
while `all_of` uses **and semantics**, rejecting entities who do not meet every specified filter.

```rust
fn 

```

> We can filter on multiple types of relations at once by chaining together our `.filter_relation` methods before we call `.build()`.

Reviewer: The `or/and` semantics of multiple filter relations have to be clarified.
```rust
query.filter_relation(a).filter_relation(b).build()
```
Is it "return everything that matches either `a` or `b`" or is it "return everything that matches both `a` and `b` at the same time"?

```rust
fn 

```

> Chaining filters in this way uses "and" semantics, just like `all_of`.

Reviewer: ````suggestion
Chaining filters in this way uses "and" semantics, just like `all_of`. Following query will only return entities that match both `a` and `b` relation filters at the same time.
```rust
query.filter_relation(a).filter_relation(b).build()
```

````

Reviewer 2: I'm planning to add more in depth examples here.

While relations filters are reset each time you leave a system, you may want to reset them sooner,

> commonly when performing multiple subqueries in bevy. To do so, simply call `query.reset_filter::<R>()` or `query.reset_filters()`.

Reviewer: Why do filters need a reset? What kind of state do they hold? I feel like starting fresh iterator on the query should always give me the same thing.

I presume the idea here is to "remove previously added relation filters". In that case, how about `query.remove_relation_filter::<R>()` and `query.clear_relation_filters()`?

Also this remove method accepts a relation type and nothing else. Does that mean I can only add a single filter per given relation type in the first place? Or does it remove all added filters with this relation type?

```rust

```

Finally, for when you have *truly* complex relation filtering needs, you can turn to **compound relation filters**.
By combining `RelationFilter::not` with the power of `RelationFilter::inner_join`, `RelationFilter::left_join`, `RelationFilter::right_join`, `RelationFilter::full_join`, you can extract the information you desire from even the most convoluted of architectures!
As you might expect, these are `(RelationFilter, RelationFilter) -> RelationFilter` functions,
and act like the [standard database joins](http://www.sql-join.com/sql-join-types) of the same names.

> No example is given: if you need these, you will know what to do.

Reviewer: Is this a part of `min-relations`? Also at the point when we do actually specify them, there should be an example given. But I think this should be only lightly mentioned under future work right now.

Reviewer 2: This will be cleaned up shortly to a much simpler paradigm thanks to Boxy.

As with all things, we can gradually improve this over time.

The set of features expressed here is useful enough to be a clear improvement over the existing patterns in at least some use cases.
There's a lot of value in getting this feature in front of users to see the response at scale.

### Why is the implementation for this so *slow*?

Many uses of relations are not performance-critical,
and the value of a nice API and automatic handling of data synchronization far outweighs any performance cost for those users.

We can fix the perf once we have a better understanding of the APIs we need to support.
There's also nothing stopping you from doing things the old ways in your elaborate graph-traversing systems.

See also: the question directly above.

## Prior art

[`Flecs`](https://github.com/SanderMertens/flecs), an advanced C++ ECS framework, has a similar feature, which they call "relationships".
These are somewhat different, they use an elaborate query domain-specific language along with being more feature rich.
You read more about them in the [corresponding PR](https://github.com/SanderMertens/flecs/pull/358).

## Unresolved questions

1. How do we put relations in Bundles?

> 2. What do we actually call [`Noitaler`][[bikeshed]? Candidates so far are:

Reviewer: ```suggestion
2. What do we actually call [`Noitaler`](#bikeshed)? Candidates so far are:
```

   1. `ReverseRelation`
   2. `Inverse`

>    3. `CoRelation`

Reviewer: I know this isn't exactly the time to bikeshed, but I think we should remove this one. It's too similar to "corelation", which means totally diffrent thing.

>    4. `TargetOf`

Reviewer: ```suggestion
   4. `TargetOf`
   5. 'ConverseRelation`
```

>    1. `ReverseRelation`
>    2. `Inverse`
>    3. `CoRelation`
>    4. `TargetOf`

Reviewer: I know this isn't exactly the time to bikeshed, but I think we should remove "CoRelation". It's too similar to "corelation", which means totally diffrent thing.

```suggestion
   1. `ReverseRelation`
   2. `Inverse`
   3. `ConverseRelation`
   4. `TargetOf`
```

Reviewer 2: IMO this is better as a separate option. "Co-relation" is the obvious term for readers with a background in advanced mathematics, even if they are a tiny minority.

Reviewer: I mean, I don't have anything strong to say here. In fact i've made a typo there, I wanted to point out similarity to "correlation" :P If you are not convinced to remove or replace this potential name, ignore this comment.

Reviewer 2: > I wanted to point out similarity to "correlation" :P

Yeah, this is the main reason I dislike CoRelation and don't think it's adequate :(. Capitalizing it like `CoRelation` is better than `Corelation` for sure, but it doesn't help enough.

Reviewer 3: `SharedRelation` seems equivalent in meaning to my untrained eyes (co -> shared) and seems to match the way I would describe this if I was teaching it to someone.

Proposed API: `Relation<T: Component, Q: WorldQuery>`
2. Graph shape guarantees (e.g tree, acyclic, 1-depth, non-self-referential).
Likely implemented using archetype invariants.
3. Graph traversals API: breadth-first, depth-first, root of tree etc.
4. Arbitrary target types instead of just `Entity` combined with `KindedEntity` proposal to ensure your targets are valid.
5. Automatically symmetric (or anti-symmetric) relations.
6. `Noitaler`*, for relations that point in the opposite direction.
7. Streaming iters to allow for queries like: `Query<&mut Money, Relation<Owes, &mut Money>>` and potentially use invariants on graph shape to allow for skipping soundness checks for the aliasing `&mut Money`'s
8. Assorted Performance optimizations.
9. Relative ordering between relations of the same kind on the same entity.
This would enable the `Styles` proposal from #1 to use relations.
10. Generalized `despawn_recursive` by parameterizing on relation type.

Relation applications in the engine:

1. Replace existing parent-child API.
2. Applications for UI: styling, widget wiring etc.
3. Frustum culling with multiple cameras.

> [bikeshed]: AUTHOR'S NOTE: Much like `yeet`, `Noitaler` is the bikeshed-avoidance name for the inverse of `Relation`, and will never be used.

Reviewer: ```suggestion
<b id="bikeshed">AUTHOR'S NOTE</b>: Much like `yeet`, `Noitaler` is the bikeshed-avoidance name for the inverse of `Relation`, and will never be used.

```

    for (source_mass, spring) in query.iter_mut(){        
        for (target_mass, _) in spring {
            assert!(source_mass != target_mass, 
                "Springs should not connect a mass to itself!");

            // Spring relations are symmetric; each spring is defined by
            // two identical relations that point opposite direction.
            // We need to preserve both during this operation
            if m1 == source_mass {
                // If mass 1 is the source, we must change the source to mass 2
                commands.entity(m1).move_relation::<Spring>(m2, target_mass);
            } else if target_mass == m1 {
                // If mass 1 is the target, we must change the target to mass 2
                commands.entity(source_mass).change_target::<Spring>(m1, m2);
            }
        }
    }
}

```

### Advanced relation filters

For those cases where previous examples weren't enough, there are several additional API extensions that you might want to use. Using those, Relation filters can be combined in arbitrarily complex ways.

The first of these is the ability to filter by the source entity as well:

```rust

> fn 

Reviewer 4: ```suggestion
// TODO complete this
fn 
```

    query: Query<(Entity, &NewOwner, Option<&Relation<Owns>>), With<Kitten>>,
) {
   for (kitten, new_owner, previous_ownership) in query.iter(){
       // Changing the target or the source will fail silently if no appropriate relation exists
       match previous_ownership {
           // We can change sources by controlling which entity owns the relation
           // move_relation is directly analagous to move_component
           Some(_, old_owner) => commands.entity(old_owner).move_relation::<Owns>(new_owner, kitten); // uwu :3
           None => commands.entity(new_owner).insert_relation(Owns::default(), kitten); // uwu!!
       }
   } 
}

// `change_target` and `move_relation` preserve the relation's data
// This system removes all springs attached to mass 1, and adds them to mass 2 instead
fn reattach_springs(mut commands: Commands,
                   selected_mass1: Res<FirstSelected>,
                   selected_mass2: Res<SecondSelected>,
                   query: Query<(Entity, Relation<Spring>)>
                  ) {
    
    let m1 = selelected_mass1.entity;
    let m2 = selelected_mass2.entity;

>     for (source_mass, spring) in query.iter_mut(){        

Reviewer: this example could use `query.get_mut(source_mass)` to avoid the outside loop. We know which entity's relations we are interested in directly.

# Feature Name: `min-relations`

## Summary

Relations allow users to create logical connection between entities. Each source entity can have multiple kinds of relations, each with their own data, and multiple relations of the same kind pointing to a different target entity.
Queries are expanded to allow efficient filtering of entities by the relations they have or the targets of these relations.

This RFC covers the minimal version of this feature, allowing for basic dynamic entity grouping and simple game logic extensions.

## Motivation

Entities often need to be aware of each other in complex ways: an attack that targets another unit, groups of entities that are controlled en-masse, or complex parent-child hierarchies that need to move together.

We *could* simply store an `Entity` in a component, then use `query.get(my_entity)` to access the relevant data.
(In this RFC, we call this the **Entity-in-component pattern**.)
But this quickly starts to balloon in complexity.

> - How do we ensure that these components are cleaned up when the entity they're pointing to is?

Reviewer 3: ```suggestion
- How do we ensure that these components are cleaned up when the entity they're pointing to is removed?
```

        .apply_filters();

    for source_entity in filtered.iter() {
        println!("{} is friends with Dorothy!", source_entity);
    }
}

// Or even look for some combination of targets!
fn caught_in_the_middle(
    mut query: Query<&mut Stress, With<Relation<FriendsWith>>>,
    dorothy: Res<Dorothy>,
    elphaba: Res<Elphaba>,
) {
    query
        // Note that we can set relation filters even if the relation is in a query filter
        .filter_relation(
            RelationFilter::<FriendsWith>::all_of()
                .target(dorothy.entity)
                .target(elphaba.entity),
        )
        .apply_filters()
        // apply_filters() returns `&mut Query` so we can just iterate straight away
        .for_each(|mut stress| {
            println!("{} is friends with Dorothy and Elphaba!", source_entity);
            stress.val = 1000;
        })
}

```

```rust
// You can use the `Entity` target returned by your relations to fetch data from the 
// target's components by combining it with `query::get()` or `query::get_component<C>()`.
fn debts_come_due(
    mut debt_query: Query<(Entity, &Relation<Owes>)>,
    mut money_query: Query<&mut Money>,
) {

>     for (debtor, debt) in debt.query() {

Reviewer 5: Sorry, meant to do this as a review.
```suggestion
    for (debtor, debt) in debt_query.iter() {
```

        // with even one Consumes relation that targets an animal
        .filter_relation::<Consumes, InFilter<{ 1 }>>(RelationFilter::any_of().targets(animals))
        .for_each(|entity| commands.entity(entity).insert(Herbivore));
}

```

### Grouping entities

By targeting a common entity, relations work fantastically as an ergonomic way to group entities.
This pattern acts as a nice complement to marker components when either:

1. You want to define the *kind* of group, rather than just membership.
2. The number of different groups you need is unknowable at compile time.

Here are some concrete examples where this pattern works well:

- you're storing multiple parts (such as meshes or sprites with their own `Transform`s) of the same in-game-object across multiple entities
- you want to dynamically group units together, like in a real-time-strategy game
- the effects created by a single entity, like the bullets they're spewing out
- you want a way to mark the controller of a unit in a queryable way, but the number of possible controllers is unknown at compile time
- you have multiple cameras, and want to mark which frustum each entity is in so they can be culled independently

> 

Reviewer 2: Per feedback, this needs a simple worked example to show users what this looks like in practice :p

TODO: Boxy explains perf implications

### Why don't we just store Entities in components?

See [Motivation].

### Why don't we just store graph data structures in a Resource?

See [Motivation] *and* [Entity Graphs].

### Why do we need arbitrarily complex relation filters??

Relation filters take advantage of archetypes to be much faster and more ergonomic than doing the same operation by naive iteration.
In order to ensure that it can always be used effectively, we need a genuinely complete DSL,
in addition to the convenient methods discussed first.

The API for compound relation filters is simple to implement and easily ignored.

### Why is my favorite feature relegated to Future Work?

It's already many thousands of words long, and the associated PR has thousands of lines of code.
Done is better than perfect, and we had to draw the line somewhere.
As with all things, we can gradually improve this over time.

The set of features expressed here is useful enough to be a clear improvement over the existing patterns in at least some use cases.
There's a lot of value in getting this feature in front of users to see the response at scale.

> ### Why is the implementation for this so *slow*?

Reviewer 3: It could be helpful to provide some measure of how slow this is, and what operations are particularly bad.

- Are queries fast? 
- Is it only adding/changing relations that is slow?

A starting point could be to run the ECS benchmark as-is with some relation filters on to see if there is a tangible impact. Perhaps seeing how adding/removing relations compares to component add/removes. Having *some* information, even if it isn't complete or perfect, would help users understand the order of magnitude they should expect.

> Many uses of relations are not performance-critical,

I only say this because "performance-critical" is a hugely relative metric.

Likely implemented using archetype invariants.
4. Graph traversals API: breadth-first, depth-first, root of tree etc.
5. Arbitrary target types instead of just `Entity` combined with `KindedEntity` proposal to ensure your targets are valid.
6. Automatically symmetric (or anti-symmetric) relations to model undirected edges.
7. `Noitaler`*, for relations that point in the opposite direction.
8. Streaming iters to allow for queries like: `Query<&mut Money, Relation<Owes, &mut Money>>` and potentially use invariants on graph shape to allow for skipping soundness checks for the aliasing `&mut Money`'s
9. Assorted performance optimizations. For example:
   1. Reducing the cost of having many archetypes.
   2. Relation storage type options to cause archetype fragmentation based on whether *any* relations of that type are present.
   3. Index-backed relations look-up.
10. Relative ordering between relations of the same kind on the same entity.
This would enable the `Styles` proposal from #1 to use relations.
11. Compound relation filters, letting you nest logic arbitrarily deep. These are just sugar / perf for very advanced users, and can wait for a while.

> 12. \[Controversial\] A full graph constraint solver DSL ala [Flecs](https://github.com/SanderMertens/flecs) for advanced querying.

Reviewer 2: ```suggestion
12. \[Controversial\] A full graph constraint solver DSL ala [Flecs](https://github.com/SanderMertens/flecs) for advanced querying.
13. Entity-to-entity event channels using relations (blocked on [bevy #2116](https://github.com/bevyengine/bevy/pull/2116)
```

Done is better than perfect, and we had to draw the line somewhere.
As with all things, we can gradually improve this over time.

The set of features expressed here is useful enough to be a clear improvement over the existing patterns in at least some use cases.
There's a lot of value in getting this feature in front of users to see the response at scale.

### Why is the implementation for this so *slow*?

Many uses of relations are not performance-critical,
and the value of a nice API and automatic handling of data synchronization far outweighs any performance cost for those users.

We can fix the perf once we have a better understanding of the APIs we need to support,
and a tangible implementation to benchmark against.
There's also nothing stopping you from doing things the old ways in your elaborate graph-traversing systems.

See also: the question directly above.

## Prior art

[`Flecs`](https://github.com/SanderMertens/flecs), an advanced C++ ECS framework, has a similar feature, which they call "relationships".
These are somewhat different, they use an elaborate query domain-specific language along with being more feature rich.

> You read more about them in the [corresponding PR](https://github.com/SanderMertens/flecs/pull/358).

Reviewer 2: ```suggestion
You can read more about them in the [docs](https://github.com/SanderMertens/flecs/blob/master/docs/Queries.md) or [corresponding PR](https://github.com/SanderMertens/flecs/pull/358).
```

       match previous_ownership {
           // We can change sources by controlling which entity owns the relation
           // move_relation is directly analagous to move_component

>            Some(old_owner, _) => commands.entity(old_owner).move_relation::<Owns>(new_owner, kitten); // uwu :3

Reviewer: ```suggestion
           Some((old_owner, _)) => commands.entity(old_owner).move_relation::<Owns>(new_owner, kitten); // uwu :3
```

### Advanced relation filters

The examples shown above show of the basics of relation filtering, but that's not all they can do.
The most important extension is the ability to filter for multiple entities at once from a `Vec<Entity>`:

```rust
// In this example, possible movess are stored on each piece
// as a relation to a specific board position
fn next_move(mut commands: Commands,
             board_state: Res<BoardState>,
             query: Query<(Entity, &Relation<Move>), With<Piece>>){    
    
    let valid_positions: Vec<Entity> = board_state.compute_valid_positions();

    let filtered = query.filter_relation::<Action, _>( 
        RelationFilter::any_of().targets(valid_positions)
    ).build();

    for (piece, valid_moves) in filtered.iter(){

>         for (potential_position, current_move) in valid_moves {
>             // Our heuristic is always >= 0
>             let mut current_best = -1.0;
>             // This is just a dummy to be replaced
>             let best_move = Entity::new();
>             let current = current_move.compute_heuristic()
> 
>             if current > current_best{
>                 current_best = current;
>                 best_move = current_move;
>             }
>         }
>         commands.insert_relation::<NextMove>(piece, best_move);

Reviewer: this `for` loop could be replaced with simple `.max_by_key(..)`;
```suggestion
		let best_move = valid_moves
			.max_by_key(|(potential_position, _)| potential_position.compute_heuristic());
			
		if let Some((_, best_move)) = best_move {
			commands.insert_relation::<NextMove>(piece, best_move);
		}
```

also changed `compute_heuristic` to be a method of `potential_position`, as it made more sense to me and also uses both tuple arguments. Please correct me if that's wrong for some reason.

Reviewer 2: Awesome. I agree with all of those changes.

Reviewer 2: Applied now.

Reviewer 2: This change has been applied <3

To do so, we set our relation filter on a `Without<Relation<R>>` query parameter.

```rust

> // Fraternizing with communists is a strict disqualifcation!

Reviewer 4: ```suggestion
// Fraternizing with communists is a strict disqualification!
```

This advanced technique is helpful when you really need to be specific, as shown in these examples:

```rust
fn herbivores(
    commands: mut Commands,
    plants_query: Query<Entity, With<Plant>>,
    animals_query: Query<Enity, With<Animal>>,
    species_query: Query<Entity, (With<Relation<Consumes>, Without<Relation<Consumes>>)>
){
    let plants = plants_query.iter().collect();
    let animals = animals_query.iter().collect();

    species_query
        // The second type parameter controls which relation filter of that type is being modified

>         .filter_relation::<Consumes, InFilter<{1}>>(RelationFilter::any_of().targets(plants))

Reviewer: I'm not sure about this one. Adding extra generic param to that function call will always make it required to list, even if this is unambiguous, e.g. like thtat
```rust
        .filter_relation::<Owes, _>>(target)
```
which just doesn't look good. Preferably we would structure it such that this generic can be auto-inferred somewhere, which should make the compiler solve the unambiguous case (given relation type is used only once in the query) without any extra help.

I know it's possible, because that's how `frunk` crate infers the `Indices` to operate on. You don't have to do anything if there is only one valid and provable set of indices for given operation. But it might require some extra gymnastics around generics for the relation filter itself or something. Also it would be nice if the index could be specified at runtime, but this is secondary.

Reviewer: Alternative syntax that avoid `_`:
```rust
        .filter_relation::(RelationFilter::<Owes>::target(target))
```

Reviewer 2: This second type parameter is definitely my largest complaint with the current design as well.
